qutip/solver/_solverqevo.pyx

```diff
@@ -0,0 +1,116 @@
+cimport qutip.core.data as _data
+from qutip.core.data.reshape cimport column_unstack_dense, column_stack_dense
+import qutip.core.data as data
+from qutip import Qobj, spre
+cimport cython
+from qutip.solver._feedback cimport (QobjFeedback, ExpectFeedback,
+                                     CollapseFeedback, Feedback)
+from qutip.core.data.base import idxint_dtype
+from libc.math cimport round
+import numpy as np
+
+cdef class SolverQEvo:
+    def __init__(self, base, options, dict args, dict feedback):
+        self.base_py = base
+        self.base = base.compiled_qobjevo
+        self.set_feedback(feedback, args, base.cte.issuper,
+                          options.ode['feedback_normalize'])
+        self.num_calls = 0
+
+    def jac_np_vec(self, t, vec):
+        return self.jac_data(t).to_array()
+
+    cdef _data.Data jac_data(self, double t):
+        if self.has_dynamic_args:
+            raise NotImplementedError("jacobian not available with feedback")
+        self.num_calls += 1
+        return self.base.call(t, data=True)
+
+    cpdef list get_coeff(self, double t, vec=None):
+        # TODO: Is this needed? I used it in debug but does it have any other use?
+        cdef Feedback feedback
+        cdef _data.Dense state
+        if self.has_dynamic_args and vec is not None:
+            state = _data.dense.Dense(vec, copy=False)
+            for dyn_args in self.dynamic_arguments:
+                feedback = <Feedback> dyn_args
+                val = feedback._call(t, state)
+                self.args[feedback.key] = val
+            for i in range(self.base.n_ops):
+                (<Coefficient> self.base.coeff[i]).arguments(self.args)
+        out = []
+        for i in range(self.base.n_ops):
+            out.append((<Coefficient> self.base.coeff[i])(t))
+        return out
+
+    def mul_np_double_vec(self, t, vec):
+        vec_cplx = vec.view(complex)
+        return self.mul_np_vec(t, vec_cplx).view(np.float64)
+
+    def mul_np_vec(self, t, vec):
+        cdef int i, row, col
+        cdef _data.Dense state = _data.dense.fast_from_numpy(vec)
+        column_unstack_dense(state, self.base.shape[1], inplace=True)
+        cdef _data.Dense out = _data.dense.zeros(state.shape[0],
+                                state.shape[1], state.fortran)
+        out = self.mul_dense(t, state, out)
+        column_stack_dense(out, inplace=True)
+        return out.as_ndarray().ravel()
+
+    cdef _data.Data mul_data(self, double t, _data.Data vec):
+        if self.has_dynamic_args:
+            self.apply_feedback(t, vec)
+        self.num_calls += 1
+        return self.base.matmul(t, vec)
+
+    cdef _data.Dense mul_dense(self, double t, _data.Dense vec, _data.Dense out):
+        if self.has_dynamic_args:
+            self.apply_feedback(t, vec)
+        self.num_calls += 1
+        return self.base.matmul_dense(t, vec, out)
+
+    def set_feedback(self, dict feedback, dict args, bint issuper, bint norm):
```

Comment:
The old name was `dynamic_arguments`, but I feel `feedback` to be more descriptive. Which name you prefer?

In previous version the function of a time-dependant system could be f(t, state, args), where the solver would feed the state as a 1D colunm-stacked numpy array. This was causing confusion so I added reserved args key: if `"state_qobj"` was a key in the `args` dict, it would take the value of the state as a Qobj, etc. So "dynamic_args", but now that `QobjEvo` is user facing and independent from the [...] solve function, it need to move out.

I did not plan to make `Feedback` user facing and use it to hide the internal state of the solver. If by making their own, you mean `feedback=[QobjFeedback(...)]`, then you need to remember that call instead of a keyword. Feels the same to me. 
If you means having the user create their feedback function, that is what I am trying to avoid. I want the user to have access to a `Qobj` and not a reshaped data object. From that `Qobj`, the user can extract the needed information using a custom function.

If we want to keep it as simple as possible, we can remove `ExpectFeedback` (which is only a shortcut to computing expectation value from the `Qobj`) and have feedback being a string of the key from args to overwrite by the solver with the state. `mcsolve` would have 2 such input, one for the state and a second for the collapse.

Reply:
I'm with Eric in that I don't really see a sensible way for a user to define their own feedback function.  The solvers have very different internals and mechanisms of action, so there's no really meaningful `Feedback` object a user could create that could be reused between different solvers.  That would be the job of `Coefficient` (which already exists) - it takes stuff the solvers have generated, and then can apply arbitrary actions on them (albeit right now it's limited to stuff that can be a Python one-liner if you want it compiled).

I'd imagine you actually call them with a slight broadening of the `args` dictionary to mean what variables should be in scope for time-dependences and how to generate them: the `args` dictionary passed to solvers like `mesolve` could be
```python
args = {
    variable_name: <method_of_generating>,
    ...
}
```

So the key is always a variable name, and the value tells you how to generate it, which is one of:
 - constant (value is purely numeric - matches existing behaviour)
 - expectation: value is a 2-tuple `('expect', operator)` where `operator` is some `Qobj` (or even `QobjEvo`)
 - state: value is a tuple beginning with the string `'state'`, and possibly a second element that represents the type/form of the state? `('state', 'dm')`
 - collapse: value is a 1-tuple `('collapse',)`

There's always going to need to be some way of passing on the information to the solvers that they should generate something, but I think I think "magic values" is a better approach to "magic keys" in the `args` dictionary (the latter is the case now), because 
 - "magic keys" don't have an obvious value you should give them when passing the dictionary
 - "magic keys" prevent the user from having meaningful variable names in their coefficients - it's not clear what `expect_0` will be filled with (`e_ops[0]`, but you can't tell what operator that is just by looking at the coefficient), but `expect_sigmaz` is semantically clear no matter what
 
 For example
 ```python
 args = {
     'trap_frequency': 2*np.pi * 20e3,
     'expect_z': ('expect', qutip.sigmaz()),
     'current_state': ('state', 'dm'),
     'which_collapse': ('collapse',),
}
result = qutip.mcsolve(H, rho0, times, args=args)
```

That way `args` means "scope" more (but changing the name is probably just annoying to people).

Reply:
^ Though with this way, I still don't really see a need for `SolverQEvo` at all.  It would all just be handled by regular `QobjEvo`, and the solvers would just have one extra function call to add stuff to the `scope` Python dictionary before calling `QobjEvo(t, args)`.